### ISSUE TITLE
Movement decrease against floors with pitch move is annoying

### ISSUE THREAD

**Issue by LoneWolfHT** [minetest/minetest#9316]
## Problem

You're working on an underwater build/transforming a sea floor. Then let's say you want to move forwards a bit: You press W but find you hardly move forwards at all since you're looking downwards towards the floor you're working on, which results in you being 'slowed down' by the floor, something that doesn't make sense considering the fact that you're technically walking on it

I've been annoyed by this to the point of turning pitch move off because it messes me up completely if I step in even a single node of water while I'm building something

## Solutions

* Removing the movement decrease for floors

## Alternatives

* Allowing mods to control this somehow

## Additional context

This is the 'movement decrease' I'm talking about https://youtu.be/sP9E4RTSyW0
(I included a roof example but I guess that may make more sense than the decrease for the floor)

**Comment by ClobberXD** [comment]
I agree that this is very annoying. However, a decrease in speed is realistic, and maintaining the same speed when there's a collision is somewhat unrealistic and "un-immersive", IMO. OTOH, I think that the deceleration is too extreme - the player is slowed down to a crawling speed, even if the player is *almost* parallel to the surface they're colliding with. Maybe a better solution would be to be more gently decelerate players?

**Comment by LoneWolfHT** [comment]
> However, a decrease in speed is realistic

My argument is that it isn't very realistic, since your feet are touching the floor, you aren't sliding along the side of a wall. I think it would make more sense to assume the player wants to walk forwards instead of into the floor (something I've never tried to do)

**Comment by ClobberXD** [comment]
> since your feet are touching the floor, you aren't sliding along the side of a wall

Good point. Colliding with the floor isn't the same as colliding with a wall.

Perhaps we could reduce deceleration when colliding in the direction of gravity? Reduce, not disable, because walking underwater is still pretty slow compared to swimming, unless Sam can't swim for nuts. :)

(Just throwing in some thoughts)

**Comment by LoneWolfHT** [comment]
> Perhaps we could reduce deceleration when colliding in the direction of gravity? Reduce, not disable, because walking underwater is still pretty slow compared to swimming, unless Sam can't swim for nuts. :)

When you're above water and flying you probably won't want that though. I would just take from the water's viscosity to determine the speed the player moves at

**Comment by ClobberXD** [comment]
> When you're not in water and flying you probably won't want that though

Surely we'd first check whether the player is in a liquid before applying this...

**Comment by LoneWolfHT** [comment]
Ok, I think I see what you mean now, it was getting a little late last night I guess.

I think decreasing movement speed at all would be a little too much. It would probably be best to see how removing the decrease completely feels first.
I'm not even sure there would be that much of a decrease in real life anyway. Since you'll still be using your arms to propel yourself

**Comment by paramat** [comment]
Hmm, the current behaviour is actually basically correct. With pitch move enabled, when colliding with an object (the seafloor) the horizontal speed should be the horizontal component of swimming speed.
This is just like walking into a wall, the result is a small sideways speed, dependent on angle.

However, in testing, the speed reduction seems a little higher than expected at moderate downwards view angles. Maybe the code needs a check, there may be a simple fix.
> Removing the movement decrease for floors

-1 it is correct and consistent behaviour, just needs tuning or fixing @random-geek 
> Allowing mods to control this somehow

-1 unnecessary complexity, better to just choose a good behaviour.
>  I think it would make more sense to assume the player wants to walk forwards
-1 when the player is in 'pitch move' mode, this is a mode for swimming in view direction, not walking underwater in normal mode. 'W' now means down into the seafloor, not horizontal walking.

(Generally i think having pitch move apply for swimming was a mistake, and it is now opening cans of worms.)

**Comment by LoneWolfHT** [comment]
> Generally i think having pitch move apply for swimming was a mistake, and it is now opening cans of worms.

I don't. I'm glad we have this feature. It just needs a little tuning

> -1 when the player is in 'pitch move' mode, this is a mode for swimming in view direction, not walking underwater in normal mode. 'W' now means down into the seafloor, not horizontal walking.

This would only be applied when the player is scraping against a floor.
When you're in a pool swimming parallel to the bottom you don't grind your body into it all the way to your destination. You put out a hand or something out to keep your body from being ground raw while you swim towards your destination

**Comment by rubenwardy** [comment]
This seems to be expected behaviour with pitch move - you're moving towards the floor, and so a component of the velocity is cancelled out by the floor. What do you want to happen? For the velocity to be the same but directed away from the pitch of the camera?

**Comment by LoneWolfHT** [comment]
> This seems to be expected behaviour with pitch move - you're moving towards the floor, and so a component of the velocity is cancelled out by the floor. What do you want to happen? For the velocity to be the same but directed away from the pitch of the camera?

If I read that correctly, yes

**Comment by paramat** [comment]
> This would only be applied when the player is scraping against a floor.

I know :)
> When you're in a pool swimming parallel to the bottom you don't grind your body into it all the way to your destination.

This is irrelevant. In MT, when you collide with a node at a certain angle, the game engine does not assume you want to suddenly travel parallel to the node at full speed, the controls are respected.
> For the velocity to be the same but directed away from the pitch of the camera?

-1, that would be inconsistent with all other movement behaviour in MT.

But i agree with trying to tune this because the horizontal component does seen too small.
